fix: support volume topology scheduling alternatives

tallaxes opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.
-->

Fixes #2742

**Description**

Fixes volume topology scheduling to correctly handle multiple topology terms in PersistentVolumes and StorageClasses.

This PR updates volume topology handling to preserve OR semantics across PV `NodeSelectorTerms` and StorageClass `AllowedTopologies`, while keeping volume-derived topology constraints separate from pod affinity.

- Return volume topology requirements as a list of scheduling alternatives instead of flattening them into one requirement set
- Process all StorageClass `AllowedTopologies`, not just the first term
- Process all PersistentVolume node affinity `NodeSelectorTerms`, not just the first term
- Compute the cross product of topology alternatives when a pod references multiple PVC-backed volumes
- Add volume requirements only to node requirements, not pod affinity, so topology spread counting uses the pod's original constraints
- Prune incompatible cross-product branches when at least one valid volume topology alternative remains
- Preserve existing behavior when every branch is incompatible; treating that case as unschedulable requires separate validation/metrics handling and will be handled independently
- Preserve hostname-only Local and HostPath PV topology terms as unconstrained alternatives after hostname affinity is ignored
- Add explicit test helper fields for multi-term PV and StorageClass topology fixtures without changing existing `Zones` helper semantics
- Add regression coverage for multi-zone PVs, multi-zone StorageClasses, and mixed valid/invalid multi-volume topology alternatives

**How was this change tested?**

* Unit tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

tallaxes committed:
fix: support multi

tallaxes committed:
fix: support multi-zone StorageClass

jmdeal commented on pkg/controllers/provisioning/scheduling/volumetopology.go:
```diff
@@ -49,17 +50,6 @@ type VolumeTopology struct {
 }
 
 func (v *VolumeTopology) Inject(ctx context.Context, pod *v1.Pod) error {
```
We'll need a rebase in light of https://github.com/kubernetes-sigs/karpenter/pull/2780, but I think the logic stays largely the same. The main difference is that rather than relying on the existing logic to iterate through different pod requirements, we'll need to add additional logic in the scheduler to iterate over volume requirements. 

Right now we iterate on pod constraints in the outer loop of the scheduler - each time a pod fails to schedule, we iterate to the next set of constraints and add it back to the queue. Now we need to consider the cross product of two sets of requirements - volume requirements and pod requirements. Given the latter is normally a small set (typically 1), I'd probably do this within the NodeClaim to keep things simple, but we could also do it in the outer loop.

tallaxes committed:
fix: adapt multi-zone volume scheduling to separated volume requirements architecture

tallaxes committed:
chore: fix linting

tallaxes commented:
Vulnerabilities flagged here and in other PRs addressed by https://github.com/kubernetes-sigs/karpenter/pull/2844

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

jmdeal commented on pkg/controllers/provisioning/scheduling/volumetopology.go:
```diff
@@ -48,30 +48,55 @@ type VolumeTopology struct {
+			continue
+		}
+
+		// Cross product: alternatives = alternatives X volAlts
+		var newAlts []scheduling.Requirements
+		for _, existing := range alternatives {
+			for _, volReq := range volAlts {
+				merged := scheduling.NewRequirements()
+				if existing != nil {
+					merged.Add(existing.Values()...)
+				}
+				merged.Add(volReq.Values()...)
+				newAlts = append(newAlts, merged)
+			}
+		}
+		alternatives = newAlts
```
There's a subtle issue here, though I don't think it will make a difference in practice. Consider a case where we have two volumes, each associated with a storage class that has two AZs in it's allowed topologies.
```yaml
allowedTopologies:
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values:
          - us-east-1a
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values:
          - us-east-1b
```
We'll end up with the following requirements when we take the cross product:
- {zone IN us-east-1a} ∩ {zone IN us-east-1a} = {zone IN us-east-1a}
- {zone IN us-east-1a} ∩ {zone IN us-east-1b} = {zone DOES_NOT_EXIST}
- {zone IN us-east-1b} ∩ {zone IN us-east-1a} = {zone DOES_NOT_EXIST}
- {zone IN us-east-1b} ∩ {zone IN us-east-1b} = {zone IN us-east-1b}

In practice this should be fine since we don't expect there to be any instance types compatible with the `topology.kubernetes.io/zone DoesNotExist` requirement. There will be some no-op iterations, but that's just a performance concern. 

You could imagine this being an issue for a label which isn't present on every instance type. Since this would result in some DoesNotExist operators, we could select an instance that doesn't have a value defined for that label when in reality it's not compatible with any instance since it has a requirement on a disjoint set of values.

I don't think there's a method which creates this exact intersection with our requirements code today. We basically want to say take the intersection for any overlapping keys and pass through any non-overlapping keys. If the overlapping keys have disjoint values, error and discard the term. Here's a contrived example I came up with:

```yaml
# Volume A SC:
allowedTopologies:
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values: [us-east-1a]
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values: [us-east-1b]

# Volume B SC:
allowedTopologies:
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values: [us-east-1a]
      - key: topology.custom.csi/rack
        values: [rack-1]
  - matchLabelExpressions:
      - key: topology.kubernetes.io/zone
        values: [us-east-1b]
      - key: topology.custom.csi/rack
        values: [rack-2]
```

- {zone IN 1a} ∩ {zone IN 1a, rack IN rack-1} = {zone IN 1a, rack IN rack-1}
- {zone IN 1a} ∩ {zone IN 1b, rack IN rack-2} = {zone DOES_NOT_EXIST, rack IN rack-2}
  - Discarded because the zonal requirements are disjoint
- {zone IN 1b} ∩ {zone IN 1a, rack IN rack-1} = {zone DOES_NOT_EXIST, rack IN rack-1}
  - Discarded because the zonal requirements are disjoint
- {zone IN 1b} ∩ {zone IN 1b, rack IN rack-2} = {zone IN 1b, rack IN rack-2}

tallaxes replied:
Yeah, good point, and disjoint alternatives could survive in the new codepath. I pulled the fix into stacked PR https://github.com/tallaxes/karpenter-core/pull/1 for a smaller review; it prunes incompatible branches before merging, threads the filtered pod set through provisioning, and adds regression coverage. @jmdeal, please let me know if this makes sense / is directionally correct, and I will merge it here. (Alternatively, could also handle this as a follow-up PR, just to land the core fix)

tallaxes replied:
As discussed in [tallaxes/karpenter-core#1](https://github.com/tallaxes/karpenter-core/pull/1#discussion_r3237176832), I pulled the mixed valid/invalid disjoint-branch handling back into this PR, in  https://github.com/kubernetes-sigs/karpenter/pull/2743/commits/2dae1125994d6ee1e68fd2b01739d1d1b1d477c2. This covers the original example by pruning incompatible cross-product branches when at least one valid volume topology alternative remains. Leaving the all-pruned case unchanged, since treating that as unschedulable needs separate validation/metrics handling in a follow-up.

tallaxes replied:
Also note a small additional fix/test in https://github.com/kubernetes-sigs/karpenter/pull/2743/commits/c922c37ddae44bdb9c4864ae07bfa2c26efd6893. It extends the multi-term OR semantics to local/hostPath PVs: when Karpenter strips hostname-keyed expressions from a PV's NodeAffinity (since it can't satisfy a specific hostname on new capacity), a term whose only expression was the hostname becomes unconstrained and, as an OR branch, makes the PV satisfiable regardless of the other terms. Added a test covering local and hostPath volumes.

I am thinking it belongs here rather than a follow-up because it's the same OR-of-NodeSelectorTerms code path being changed - just the local/hostPath flavor of it - but let me know if you disagree.

jmdeal commented on pkg/controllers/provisioning/scheduling/existingnode.go:
```diff
@@ -91,14 +91,41 @@ func (n *ExistingNode) CanAdd(pod *v1.Pod, podData *PodData, volumes scheduling.
+		volumeAlternatives = []scheduling.Requirements{nil}
+	}
+
+	// Try each volume alternative. Choosing a zone for volumes affects topology checks.
```
I've noticed quite a few comments are making the assumption that topology will only be zones. Since we support arbitrary CSI drivers, we shouldn't make that assumption. I know at least one counter-example used by Karpenter today: storage classes for EKS Auto Mode's EBS CSI Driver should be configured with a `eks.amazonaws.com/compute-type: auto` selector in allowed topologies.

tallaxes replied:
Addressed https://github.com/kubernetes-sigs/karpenter/pull/2743/commits/15e740424b3634f65338fd39307f46c4c930b844

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
test: add explicit multi-term volume topology fixtures

tallaxes committed:
docs: stop assuming volume topology is zonal

tallaxes commented on pkg/test/storage.go:
Done; added a way to pass selector terms into options https://github.com/kubernetes-sigs/karpenter/pull/2743/commits/fce197f976d3ddd90723b7e8556f1e3c76e34c80

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
Merge branch 'main' into tallaxes/muli-zone-pv-scheduling

tallaxes committed:
fix: prune disjoint volume topology alternatives

Discard incompatible branches when merging multi-volume topology alternatives, while preserving existing behavior when all branches are incompatible.

tallaxes committed:
fix: preserve volume topology alternatives

Preserve Local and HostPath PV node selector terms that become empty
after hostname affinity is ignored, since those terms should remain
valid unconstrained alternatives.

Also keep the multi-zone PV regression test using separate PV/PVC
fixtures per pod so it exercises OR semantics without sharing a single
ReadWriteOnce claim.

jmdeal commented:
/lgtm
/approve
